### PR TITLE
ComputeJacobianBlocksThread: add face contributions

### LINKED ISSUE #13886: ComputeJacobianBlocksThread ignores Jacobian elements on faces
## Bug Description

For DG kernels, there is coupling between different elements though face calculations.

`ComputeJacobianBlocksThread` ignores the face contributions that are important for DG calculations.  Reported by @jthano

## Steps to Reproduce

Use DG kernels and `ComputeJacobianBlocksThread`

## Impact

DG users

### PR DESCRIPTION
**Pull request by fdkong**
Closes #13886

### CODE REVIEW

**Comment by fdkong** [comment]
Could you try if it works for you @jthano?

File: framework/src/loops/ComputeJacobianBlocksThread.C [hunk 1]
```diff
@@ -37,17 +37,46 @@ ComputeJacobianBlocksThread::~ComputeJacobianBlocksThread() {}
...
+    dof_map.dof_indices(elem, _dof_indices);
 
     _fe_problem.addJacobianBlock(
-        block->_jacobian, block->_ivar, block->_jvar, dof_map, dof_indices, _tid);
+        block->_jacobian, block->_ivar, block->_jvar, dof_map, _dof_indices, _tid);
+  }
+}
+
+void
+ComputeJacobianBlocksThread::postInternalSide(const Elem * elem, unsigned int side)
```
Comment (YaqiWang) [inline framework/src/loops/ComputeJacobianBlocksThread.C]: This looks like not the right fix. `ComputeJacobianThread` did the adding in `onInternalSide`, `onInterface`, `postElement`. We possibly need to move the adding in `onInternalSide`, `onInterface` out to a new function there as `postInternalSide`. Then we can override `postInternalSide` here.

Comment (fdkong) [reply framework/src/loops/ComputeJacobianBlocksThread.C]: I think it is fine as is right now.

Comment (YaqiWang) [reply framework/src/loops/ComputeJacobianBlocksThread.C]: Sounds ok to me.

**Comment by fdkong** [comment]
I will add one example to test this change

File: framework/include/loops/ComputeJacobianBlocksThread.h [hunk 2]
```diff
@@ -53,6 +53,12 @@ class ComputeJacobianBlocksThread : public ComputeFullJacobianThread
 protected:
   virtual void postElement(const Elem * elem) override;
 
+  virtual void postInternalSide(const Elem * elem, unsigned int side) override;
+
   std::vector<JacobianBlock *> _blocks;
-};
 
+private:
```
Comment (lindsayad) [inline framework/include/loops/ComputeJacobianBlocksThread.h]: Look at you using `private`. I love it!

File: framework/src/loops/ComputeJacobianBlocksThread.C [hunk 3]
```diff
@@ -37,17 +37,50 @@ ComputeJacobianBlocksThread::~ComputeJacobianBlocksThread() {}
...
+  {
+    _dof_indices.clear();
+    _dof_neighbor_indices.clear();
+
+    // Pointer to the neighbor we are currently working on.
+    const Elem * neighbor = elem->neighbor_ptr(side);
+
+    Threads::spin_mutex::scoped_lock lock(Threads::spin_mtx);
+    // Get the global id of the element and the neighbor
+    const dof_id_type elem_id = elem->id(), neighbor_id = neighbor->id();
```
Comment (lindsayad) [inline framework/src/loops/ComputeJacobianBlocksThread.C]: dof_id_type -> auto

File: framework/src/loops/ComputeJacobianBlocksThread.C [hunk 4]
```diff
@@ -37,17 +37,50 @@ ComputeJacobianBlocksThread::~ComputeJacobianBlocksThread() {}
...
+
+    if ((neighbor->active() && (neighbor->level() == elem->level()) && (elem_id < neighbor_id)) ||
+        (neighbor->level() < elem->level()))
+      for (const auto & block : _blocks)
+      {
+        const DofMap & dof_map = block->_precond_system.get_dof_map();
```
Comment (lindsayad) [inline framework/src/loops/ComputeJacobianBlocksThread.C]: DofMap -> auto

**Review by lindsayad — CHANGES_REQUESTED** [review]
I'll let you decide at your discretion whether you want to make the `auto` changes. I think there's a balance between code flexibility and readability.

Can you add [Problem/error_on_jacobian_nonzero_reallocation=true] for your test?

File: test/tests/preconditioners/pbp/pbp_test.i [hunk 5]
```diff
@@ -31,6 +29,11 @@
   [../]
 []
 
+[Problem]
+  type = FEProblem
+  error_on_jacobian_nonzero_reallocation = true
```
Comment (lindsayad) [inline test/tests/preconditioners/pbp/pbp_test.i]: Could you add this to the DG version as well? I'm more worried about that one...

Comment (fdkong) [reply test/tests/preconditioners/pbp/pbp_test.i]: I was intending to add this to DG version. I looked at the wrong file :)

**Review by lindsayad — DISMISSED** [review]

**Comment by lindsayad** [comment]
Uh oh, tabs!

**Comment by fdkong** [comment]
> Uh oh, tabs!

I love tabs. I will be going to change precheck script:)

**Comment by fdkong** [comment]
@lindsayad Ready for review again

**Review by lindsayad — APPROVED** [review]

**Pull request merged**